File: 2026/05-May/30-Saturday/Solution.java
import java.util.*;

class SegmentTree {
    int n;
    int[] tree;

    SegmentTree(int n) { 
        this.n = n;
        tree = new int[4 * n];
    }

    void update(int node, int l, int r, int idx, int val) {
        if (l == r) {
            tree[node] = val;
            return;
        }

        int mid = (l + r) >> 1;

        if (idx <= mid)
            update(node * 2, l, mid, idx, val);
        else
            update(node * 2 + 1, mid + 1, r, idx, val);

        tree[node] = Math.max(tree[node * 2], tree[node * 2 + 1]);
    }

    int query(int node, int l, int r, int ql, int qr) {
        if (ql > r || qr < l)
            return 0;

        if (ql <= l && r <= qr)
            return tree[node];

        int mid = (l + r) >> 1;

        return Math.max(
            query(node * 2, l, mid, ql, qr),
            query(node * 2 + 1, mid + 1, r, ql, qr)
        );
    }
}

class Solution {
    public List<Boolean> getResults(int[][] queries) {
        int MAX = 50000;

        TreeSet<Integer> obstacles = new TreeSet<>();
        obstacles.add(0);

        SegmentTree seg = new SegmentTree(MAX + 1);

        List<Boolean> ans = new ArrayList<>();

        for (int[] q : queries) {

            if (q[0] == 1) {

                int x = q[1];

                Integer left = obstacles.floor(x);
                Integer right = obstacles.ceiling(x);

                obstacles.add(x);

                seg.update(1, 0, MAX, x, x - left);

                if (right != null) {
                    seg.update(1, 0, MAX, right, right - x);
                }

            } else {

                int x = q[1];
                int sz = q[2];

                Integer prev = obstacles.floor(x);

                int best = seg.query(1, 0, MAX, 0, prev);

                ans.add(
                    best >= sz ||
                    x - prev >= sz
                );
            }
        }

        return ans;
    }
}

File: 2026/05-May/30-Saturday/solution.py
from bisect import bisect_right
from sortedcontainers import SortedList

class FenwickTree: 
    def __init__(self, n):
        self.bit = [0] * (n + 1)

    def update(self, idx, val):
        while idx < len(self.bit):
            self.bit[idx] = max(self.bit[idx], val)
            idx += idx & -idx

    def query(self, idx):
        res = 0
        while idx > 0:
            res = max(res, self.bit[idx])
            idx -= idx & -idx
        return res


class Solution:
    def getResults(self, queries):
        limit = min(50000, len(queries) * 3)

        obstacles = SortedList([0, limit])
        bit = FenwickTree(limit + 2)

        for q in queries:
            if q[0] == 1:
                obstacles.add(q[1])

        for i in range(len(obstacles) - 1):
            left = obstacles[i]
            right = obstacles[i + 1]
            bit.update(right, right - left)

        ans = []

        for q in reversed(queries):
            if q[0] == 1:
                x = q[1]

                idx = obstacles.index(x)
                prev_ob = obstacles[idx - 1]
                next_ob = obstacles[idx + 1]

                obstacles.remove(x)
                bit.update(next_ob, next_ob - prev_ob)

            else:
                x, sz = q[1], q[2]

                idx = obstacles.bisect_right(x)
                prev_ob = obstacles[idx - 1]

                ans.append(
                    bit.query(prev_ob) >= sz or
                    x - prev_ob >= sz
                )

        return ans[::-1]
